File: app/src/main/java/com/example/milen/hope/SeeLists.java
package com.example.milen.hope;

import android.content.Intent;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.Spinner;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

public class SeeLists extends AppCompatActivity {
    private static final String TAG = "SeeListsActivity";
    private DatabaseReference mDatabaseReference;
    private DatabaseReference mItemsReference;
    GoogleSignInClient mGoogleSignInClient;
    GoogleSignInAccount account;
    private static final int RC_SIGN_IN = 4444;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_see_lists);

        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build();
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        mDatabaseReference = FirebaseDatabase.getInstance().getReference();


        Spinner listPicker = findViewById(R.id.spinner);
        listPicker.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                SeeLists.this.fetchList(parent.getItemAtPosition(position).toString());
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                Log.w(TAG, "User has not selected an item");
            }
        });

        FloatingActionButton addButton = findViewById(R.id.add_list_button);
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addButtonCLick();
            }
        });

        FloatingActionButton removeButton = findViewById(R.id.remove_list_button);
        removeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                removeButtonCLick();
            }
        });
    }

    private void removeButtonCLick() {
    }

    private void addButtonCLick() {
        Intent intent = new Intent(this, CreateList.class);
        startActivity(intent);
    }

    @Override
    public void onStart(){
        super.onStart();

        account = GoogleSignIn.getLastSignedInAccount(SeeLists.this);
        if (account == null)
            signInUser();


        ValueEventListener listsListener = new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                ArrayList<String> dropdownOptions= new ArrayList<>();
                Iterable<DataSnapshot> children = dataSnapshot.getChildren();
                for (DataSnapshot child : children){
                    dropdownOptions.add(child.getKey());
                }

                ArrayAdapter<String> adapter = new ArrayAdapter<>(SeeLists.this,
                        android.R.layout.simple_spinner_dropdown_item, dropdownOptions);
                Spinner listPicker = findViewById(R.id.spinner);
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                listPicker.setAdapter(adapter);

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
                // Getting Post failed, log a message
                Log.w(TAG, "The following dabase error occured: ", databaseError.toException());
            }
        };

        // TODO: Change this.
        while (account == null) {
            // Simply wait
        }

        DatabaseReference mListsReference = mDatabaseReference.child("users").child(account.getId());
        mListsReference.addListenerForSingleValueEvent(listsListener);
    }


    public void signInUser() {
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);

            try {
                account = task.getResult(ApiException.class);

                // Signed in successfully, show authenticated UI.
            } catch (ApiException e) {
                // The ApiException status code indicates the detailed failure reason.
                // Please refer to the GoogleSignInStatusCodes class reference for more information.
                Log.w(TAG, "signInResult:failed code=" + e.getStatusCode());
            }

        }
    }

    public void fetchList(final String listName) {

        ValueEventListener listItemsListener = new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                Iterable<DataSnapshot> children = dataSnapshot.getChildren();
                LinearLayout layout = (LinearLayout) SeeLists.this.findViewById(R.id.layout_items);
                if (layout == null)
                    return; // TODO : Create an error or something.
                if (layout.getChildCount() > 0) {
                    layout.removeAllViewsInLayout();
                }
                for (DataSnapshot child : children) {
                    CheckBox checkBox = new CheckBox(SeeLists.this);
                    checkBox.setText(child.getKey());
                    if (child.getValue().toString().equals("false"))
                        checkBox.setChecked(false);
                    else
                        checkBox.setChecked(true);
                    SeeLists.this.linkWithListener(checkBox);
                    layout.addView(checkBox);
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
                // Getting Post failed, log a message
                Log.w(TAG, "The following dabase error occured: "
                        , databaseError.toException());
            }
        };
        mItemsReference = mDatabaseReference.child("users").child(account.getId())
                .child(listName);
        mItemsReference.addValueEventListener(listItemsListener);
    }

    public void linkWithListener(CheckBox checkBox) {
        checkBox.setOnCheckedChangeListener(new CheckBox.OnCheckedChangeListener() {

            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                String changedItem = String.valueOf(buttonView.getText());
                mItemsReference.child(changedItem).setValue(isChecked);
            }
        });
    }
}

File: app/src/main/java/com/example/milen/hope/SeeLists.kt
package com.example.milen.hope

import android.content.Intent
import android.support.design.widget.FloatingActionButton
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.LinearLayout
import android.widget.Spinner

import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

import java.util.ArrayList

class SeeLists : AppCompatActivity() {
    private var mDatabaseReference: DatabaseReference? = null
    private var mItemsReference: DatabaseReference? = null
    internal var mGoogleSignInClient: GoogleSignInClient? = null
    internal var account: GoogleSignInAccount? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_see_lists)

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build()
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso)

        mDatabaseReference = FirebaseDatabase.getInstance().reference


        val listPicker = findViewById<Spinner>(R.id.spinner)
        listPicker.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                this@SeeLists.fetchList(parent.getItemAtPosition(position).toString())
            }

            override fun onNothingSelected(parent: AdapterView<*>) {
                Log.w(TAG, "User has not selected an item")
            }
        }

        val addButton = findViewById<FloatingActionButton>(R.id.add_list_button)
        addButton.setOnClickListener { addButtonCLick() }

        val removeButton = findViewById<FloatingActionButton>(R.id.remove_list_button)
        removeButton.setOnClickListener { removeButtonCLick() }
    }

    private fun removeButtonCLick() {}

    private fun addButtonCLick() {
        val intent = Intent(this, CreateList::class.java)
        startActivity(intent)
    }

    public override fun onStart() {
        super.onStart()

        account = GoogleSignIn.getLastSignedInAccount(this@SeeLists)
        if (account == null)
            signInUser()


        val listsListener = object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val dropdownOptions = ArrayList<String>()
                val children = dataSnapshot.children
                for (child in children) {
                    dropdownOptions.add(child.key!!)
                }

                val adapter = ArrayAdapter(this@SeeLists,
                        android.R.layout.simple_spinner_dropdown_item, dropdownOptions)
                val listPicker = findViewById<Spinner>(R.id.spinner)
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                listPicker.adapter = adapter

            }

            override fun onCancelled(databaseError: DatabaseError) {
                // Getting Post failed, log a message
                Log.w(TAG, "The following dabase error occured: ", databaseError.toException())
            }
        }

        // TODO: Change this.
        while (account == null) {
            // Simply wait
        }

        val mListsReference = mDatabaseReference!!.child("users").child(account!!.id!!)
        mListsReference.addListenerForSingleValueEvent(listsListener)
    }


    fun signInUser() {
        val signInIntent = mGoogleSignInClient!!.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)

            try {
                account = task.getResult<ApiException>(ApiException::class.java)

                // Signed in successfully, show authenticated UI.
            } catch (e: ApiException) {
                // The ApiException status code indicates the detailed failure reason.
                // Please refer to the GoogleSignInStatusCodes class reference for more information.
                Log.w(TAG, "signInResult:failed code=" + e.statusCode)
            }

        }
    }

    fun fetchList(listName: String) {

        val listItemsListener = object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val children = dataSnapshot.children
                val layout = this@SeeLists.findViewById<View>(R.id.layout_items) as LinearLayout
// TODO : Create an error or something.
                if (layout.childCount > 0) {
                    layout.removeAllViewsInLayout()
                }
                for (child in children) {
                    val checkBox = CheckBox(this@SeeLists)
                    checkBox.text = child.key
                    checkBox.isChecked = child.value!!.toString() != "false"
                    this@SeeLists.linkWithListener(checkBox)
                    layout.addView(checkBox)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
                // Getting Post failed, log a message
                Log.w(TAG, "The following dabase error occured: ", databaseError.toException())
            }
        }
        mItemsReference = mDatabaseReference!!.child("users").child(account!!.id!!)
                .child(listName)
        mItemsReference!!.addValueEventListener(listItemsListener)
    }

    fun linkWithListener(checkBox: CheckBox) {
        checkBox.setOnCheckedChangeListener(object : CompoundButton.OnCheckedChangeListener {

            override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
                val changedItem = buttonView.text.toString()
                mItemsReference!!.child(changedItem).setValue(isChecked)
            }
        })
    }

    companion object {
        private val TAG = "SeeListsActivity"
        private val RC_SIGN_IN = 4444
    }
}
